Cluster features on raft

piodul opened the pull request:
This PR adds support for managing cluster features using raft. This new method is more safe than the existing one as it doesn't rely on gossip to propagate information about nodes' supported features in time. It is only enabled if raft is enabled and experimental raft feature is turned on.

A new column named `supported_features` is added to the group0-managed `system.topology` table. Every node is responsible for inserting and maintaining its features in that table: they are inserted when a node inserts its request to join the cluster or replace a node, and updated when a node restarts and notices that its supported feature set is different to what is in the table. The set of _enabled_ features in the cluster is implicitly defined as an intersection of all nodes' supported features persisted in the topology table.

Until a node puts or updates its features into the topology table on startup, it doesn't have a guarantee that it will understand all of the enabled features. Therefore, a feature check is performed every time a group 0 command that modifies the topology table is applied - further commands may depend on some features, so a node shouldn't attempt to process them if it might not understand them properly.

Similarly, features are also checked on snapshot transfer. A set of enabled features is sent in a simple format along the snapshot so that a node doesn't need to attempt to parse the rest of it, which it might be not prepared to understand.

When the new cluster feature management method is enabled, the node will still advertise its supported features in gossip but it won't use gossip to judge whether a feature should be enabled or not. This is left mostly for the purpose of some logic that reacts to features in gossip (it's mostly about updating `system.peers`) and for the sake of nodes that don't use cluster features on raft.

The implementation currently assumes that the cluster either always uses raft-based features or gossip-based features. There is no logic implemented yet that allows to transition from gossip features to raft features. Such a transition will have to happen when the cluster switches to managing topology over raft - which is also not implemented, but planned, so transitioning features to the new method will have to be a part of the solution.

Along with the implementation, a bunch of pylib tests for cluster features are added. A test-only feature is added which allows simulating node upgrade to a version which supports a new feature in the tests. The test suite is being run both for the gossip-based implementation as well as the new raft-based one.

piodul commented on service/topology_state_machine.cc:
```diff
@@ -34,6 +35,49 @@ bool topology::contains(raft::server_id id) {
 }
 
+bool topology::has_complete_feature_information() const noexcept {
+    return (normal_nodes.size() + transition_nodes.size() + new_nodes.size()) == supported_features.size();
```
(Self-reviewing here)

This check is wrong. It is based on an assumption that all nodes in the cluster are already reflected in the `topology` table and, during the upgrade, the `supported_features` column of upgraded nodes will be gradually filled. However, this won't be true because, most likely, features in raft and topology changes on raft will be introduced in the same release.

kbr-scylla commented on service/topology_state_machine.hh:
```diff
@@ -83,6 +83,9 @@ struct topology {
     std::unordered_map<raft::server_id, request_param> req_param;
 
+    // Supported features, for non-left nodes
+    std::unordered_map<raft::server_id, std::set<sstring>> supported_features;
```
Have you considered putting the set inside `replica_state`?

piodul replied:
No, but makes sense to move it there.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -316,6 +316,22 @@ future<> storage_service::topology_state_load() {
+        // performing a successful read barrier makes sure that the local node
+        // supports all cluster's features.
+        const auto supported_by_me = _feature_service.supported_feature_set();
+        service::topology::check_if_enabled_features_are_known(supported_by_me, enabled_features);
```
IIUC if the check fails, this will just crash the Raft server applier fiber and report a cryptic log message but the node will continue running. We somehow need to do the check in a controlled fashion during the startup procedure and fail the startup procedure, shutting down the node if the check fails - like in the original gossiper-based code. The check should never fail once a node has started. I'm not sure even that it belongs inside `topology_state_load()`. After all this is supposed to be a deterministic state machine so either all nodes should succeed applying a command or none should.

piodul replied:
I'm not sure how this check could be implemented differently in this design.

If you wanted to do it in `join_token_ring` by taking a read barrier and then checking other nodes' enabled features, then it is probably already too late - the barrier might require us to apply some commands that the node will misunderstand or not understand at all because of the missing features. In that case the group 0 state might become wrong. Something like this can happen both in the case when the node joins the cluster or is downgraded.

kbr-scylla replied:
Maybe we could do something like this: in `join_token_ring` when we join group 0 / do read barrier on restart, first create a channel (and destroy it after read barrier). The sending side of the channel would be in `apply`/`transfer_snapshot`. If we fail the feature check in `apply`/`transfer_snapshot` we should assert that the channel is currently initialized (we should never fail the feature check after the node has started) and push the error through the channel, so that the startup procedure can fail.

piodul replied:
It appears that the exception is propagated properly from `raft::server` operations after the applier fiber stopped:

```
ERROR 2023-05-16 13:51:40,699 [shard 0] init - Startup failed: raft::stopped_error (Raft instance is stopped, reason: "background error, std::_Nested_exception<raft::state_machine_error> (State machine error at raft/raft.hh:299): std::runtime_error (Feature check failed. The node doesn't support some of the features that are enabled in the cluster. Unsupported features = {TEST_ONLY_FEATURE}, supported features = {AGGREGATE_STORAGE_OPTIONS, ALTERNATOR_STREAMS, ALTERNATOR_TTL, CDC, CDC_GENERATIONS_V2, COLLECTION_INDEXING, COMPUTED_COLUMNS, CORRECT_COUNTER_ORDER, CORRECT_IDX_TOKEN_IN_SECONDARY_INDEX, CORRECT_NON_COMPOUND_RANGE_TOMBSTONES, CORRECT_STATIC_COMPACT_IN_MC, COUNTERS, DIGEST_FOR_NULL_VALUES, DIGEST_INSENSITIVE_TO_EXPIRY, DIGEST_MULTIPARTITION_READ, EMPTY_REPLICA_PAGES, HINTED_HANDOFF_SEPARATE_CONNECTION, INDEXES, KEYSPACE_STORAGE_OPTIONS, LARGE_COLLECTION_DETECTION, LARGE_PARTITIONS, LA_SSTABLE_FORMAT, LWT, MATERIALIZED_VIEWS, MC_SSTABLE_FORMAT, MD_SSTABLE_FORMAT, ME_SSTABLE_FORMAT, NONFROZEN_UDTS, PARALLELIZED_AGGREGATION, PER_TABLE_CACHING, PER_TABLE_PARTITIONERS, RANGE_SCAN_DATA_VARIANT, RANGE_TOMBSTONES, ROLES, ROW_LEVEL_REPAIR, SCHEMA_COMMITLOG, SCHEMA_TABLES_V3, SECONDARY_INDEXES_ON_STATIC_COLUMNS, SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT, STREAM_WITH_RPC_STREAM, SUPPORTS_RAFT_CLUSTER_MANAGEMENT, TOMBSTONE_GC_OPTIONS, TRUNCATION_TABLE, TYPED_ERRORS_IN_READ_RPC, UDA, UDA_NATIVE_PARALLELIZED_AGGREGATION, UDF, UNBOUNDED_RANGE_TOMBSTONES, VIEW_VIRTUAL_COLUMNS, WRITE_FAILURE_REPLY, XXHASH}, enabled features in the cluster = {AGGREGATE_STORAGE_OPTIONS, ALTERNATOR_STREAMS, ALTERNATOR_TTL, CDC, CDC_GENERATIONS_V2, COLLECTION_INDEXING, COMPUTED_COLUMNS, CORRECT_COUNTER_ORDER, CORRECT_IDX_TOKEN_IN_SECONDARY_INDEX, CORRECT_NON_COMPOUND_RANGE_TOMBSTONES, CORRECT_STATIC_COMPACT_IN_MC, COUNTERS, DIGEST_FOR_NULL_VALUES, DIGEST_INSENSITIVE_TO_EXPIRY, DIGEST_MULTIPARTITION_READ, EMPTY_REPLICA_PAGES, HINTED_HANDOFF_SEPARATE_CONNECTION, INDEXES, KEYSPACE_STORAGE_OPTIONS, LARGE_COLLECTION_DETECTION, LARGE_PARTITIONS, LA_SSTABLE_FORMAT, LWT, MATERIALIZED_VIEWS, MC_SSTABLE_FORMAT, MD_SSTABLE_FORMAT, ME_SSTABLE_FORMAT, NONFROZEN_UDTS, PARALLELIZED_AGGREGATION, PER_TABLE_CACHING, PER_TABLE_PARTITIONERS, RANGE_SCAN_DATA_VARIANT, RANGE_TOMBSTONES, ROLES, ROW_LEVEL_REPAIR, SCHEMA_COMMITLOG, SCHEMA_TABLES_V3, SECONDARY_INDEXES_ON_STATIC_COLUMNS, SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT, STREAM_WITH_RPC_STREAM, SUPPORTS_RAFT_CLUSTER_MANAGEMENT, TEST_ONLY_FEATURE, TOMBSTONE_GC_OPTIONS, TRUNCATION_TABLE, TYPED_ERRORS_IN_READ_RPC, UDA, UDA_NATIVE_PARALLELIZED_AGGREGATION, UDF, UNBOUNDED_RANGE_TOMBSTONES, VIEW_VIRTUAL_COLUMNS, WRITE_FAILURE_REPLY, XXHASH})")
```

There is a lengthy prefix indicating several layers of exception nesting, but the error message is clearly preserved. I'm not sure whether it's worth adding a channel.

> we should never fail the feature check after the node has started

This should be implicitly true - after the features of a node are persisted in the `topology` table, it is guaranteed that no features unknown to that node will be enabled (at least until the node is removed from the cluster).

Of course, if we have a bug then the above might not be true, but then we are in a pickle. If we choose to do the check anyway and stop the node if it fails, then we risk an outage but no bad things like "one node misinterprets commands from another node" will happen. On the other hand, if the node continues to work despite the failing check, there is a risk of bad things happening but we might get away with it, and we will have better availability.

I'm not sure which is better, but I think the second variant is more similar to our current behavior, so I'll go with it for the time being. Perhaps we can print an error message to logs if we detect such an issue.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -316,6 +316,22 @@ future<> storage_service::topology_state_load() {
+        const auto supported_by_me = _feature_service.supported_feature_set();
+        service::topology::check_if_enabled_features_are_known(supported_by_me, enabled_features);
+        std::set<std::string_view> enabled_features_v = boost::copy_range<std::set<std::string_view>>(enabled_features);
+        co_await _feature_service.enable(std::move(enabled_features_v));
```
We should optimize this somehow so we don't reenable all features on every topology commands, which involves disk writes and a bunch of log messages which are just going to spam the logs unnecessarily. We should enable only whatever new features have appeared. (The optimization looks like it belongs inside `feature_service` more than here though)

piodul replied:
It's not that bad, if a feature is already enabled then `_feature_service.enable()` won't enable them again (i.e. trigger callbacks, produce log messages or do disk writes).

The only costly thing that it does unconditionally is spawning a `seastar::async`. I guess this could be avoided if no new features are to be enabled, but I'm not sure whether it's worth it.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1019,6 +1031,43 @@ future<> storage_service::raft_bootstrap(raft::server& raft_server) {
+        const sstring features_string = gms::feature_service::from_feature_set(features_set);
+
+        auto guard = co_await _group0->client().start_operation(&_abort_source);
+        topology_mutation_builder builder(guard.write_timestamp(), raft_server.id());
```
I wonder if this is safe, since the topology coordinator might also be updating this entry concurrently (?) Consider the situation: the topology coordinator has decided to remove this node from the cluster. It removes the `topology` row for this server. However concurrently we're re-adding the row by the command we send here.

(this can only happen if we're still a member of group 0 - perhaps the coordinator, when removing a node, should start by removing it from group 0, and only then removing the `topology` entry for this node)

Perhaps after taking the group 0 guard, we should check if we're still a member of the cluster (by looking if we have an entry in the state machine). This should be done while holding the guard for atomicity.

piodul commented:
v2:

- Added tests,
- Added logic for switching off enabling features based on gossip if the new method is enabled on startup.

TODO:

- One or two commits will have to be redone after https://github.com/scylladb/scylladb/pull/13172 is merged,
- Adjust the feature check on command application to use `on_internal_error` after the node inserted/updated its features in group 0.

Won't do:

> [ ] Transitional logic that disables reacting to gossip features after the cluster is ready to support features in raft (currently, both will run in parallel).

Didn't do it and won't do it here, as the "switch from gossip features to raft features" should be a part of the "switch to topology over raft" procedure, which we don't have it yet.

> [ ] Adjust the rest of the system not to react to gossip features after the switch is complete.

Didn't do it. AFAIK the only place that depends on that is updating `system.peers` in `storage_service::on_change` which also does it for some parameters that are already managed through raft. I think it will be better to adjust that function to react to changes in `system.topology` directly, but that sounds like something that is out of scope of this PR.

piodul commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1427/

Some tests failed because they rely on error injection, and I didn't tell the testing framework to skip them in release mode. Will fix.

piodul commented:
v2.2:
- Rebased and resolved conflicts
- Listed the cluster feature tests in the `skip_in_release` section of `suite.yaml` - those tests depend on a cluster feature that is enabled via error injection so they don't work in release mode
- Applied the `reconnect_driver` driver workaround to tests as I saw a failure where the driver couldn't reconnect after node restart

piodul commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1444/

- [Tests / Unit Tests / test_tablets.test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1444/testReport/junit/(root)/test_tablets/Tests___Unit_Tests___test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode/)
  The test seems to be flaky on master, so it's probably an unrelated failure.
- [raft.Master Test Suite.debug.backpressure_prevote_drops](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1444/testReport/junit/raft.Master%20Test%20Suite/debug/backpressure_prevote_drops/)
  Not sure what went wrong here... This is a test for the low level raft layer which I didn't modify. I didn't get this error locally.

kbr-scylla commented:
> Therefore, a feature check is performed every time a group 0 command that modifies the topology table is applied - further commands may depend on some features, so a node shouldn't attempt to process them if it might not understand them properly.

I wonder if this won't conflict with command merging that @gleb-cloudius is working on
If enabling features happens through `topology_change` commands, and we merge all `topology_change` commands in a batch, and a later command depends on an earlier enabled feature...

piodul commented:
> I wonder if this won't conflict with command merging that @gleb-cloudius is working on If enabling features happens through `topology_change` commands, and we merge all `topology_change` commands in a batch, and a later command depends on an earlier enabled feature...

I saw a predicate in Gleb's code that decides whether a command should be merged in or not. If a command touches the `supported_features` column then it just should be applied separately from others. We need to be careful not to forget to update that predicate - that's either on me or Gleb, depending on whose changes get merged in first.

gleb-cloudius commented:
> > I wonder if this won't conflict with command merging that @gleb-cloudius is working on If enabling features happens through `topology_change` commands, and we merge all `topology_change` commands in a batch, and a later command depends on an earlier enabled feature...
> 
> I saw a predicate in Gleb's code that decides whether a command should be merged in or not. If a command touches the `supported_features` column then it just should be applied separately from others. We need to be careful not to forget to update that predicate - that's either on me or Gleb, depending on whose changes get merged in first.

I did not look at the patch series yet, but how is it different from getting all the mutations merged through a snapshot? Or does this series implements special handling for features into the snapshot transfer? If yes, well here we see why it is a bad idea.

piodul commented:
> I did not look at the patch series yet, but how is it different from getting all the mutations merged through a snapshot? Or does this series implements special handling for features into the snapshot transfer?

The list of features enabled at the time of the snapshot is provided in a separate field. The recipient checks whether it supports all of them before applying the mutations.

As far as I understand, the recipient shouldn't try to interpret mutations if it is not sure whether it will understand them. Sender doesn't have this problem and can compute the set of enabled features for the recipient.

> If yes, well here we see why it is a bad idea.

Could you elaborate? Is it because it complicates reasoning about whether it's OK to merge mutations? I don't think it is avoidable if we use group 0 to manage features as one command can affect how later commands are interpreted.

Maybe we could have a different variant than `topology_change` just for features to make it more explicit, but it would definitely make the rest of the implementation more complicated. For example, when a node joins, it inserts basic information about itself to the system.topology table, including its own features and the join request - it should happen in a single command.

kbr-scylla commented on test/pylib/util.py:
```diff
@@ -133,4 +133,16 @@ async def feature_is_enabled():
     await wait_for(feature_is_enabled, deadline)
```
We need a fourth test:
upgrade the cluster so it enables the feature, then try to downgrade one of the nodes. The node should refuse to start.

kbr-scylla commented on test/topology/test_cluster_features.py:
```diff
@@ -0,0 +1,163 @@
+
+
+@pytest.mark.asyncio
+async def test_reject_joining_node_without_support_for_enabled_features(manager: ManagerClient) -> None:
```
Can we merge this and below test? I don't think there's any value in bootstrapping a completely fresh cluster - it will only prolong CI runs etc. 

We can try the bootstrap, see it fail, and then try the replace in a single test.

gleb-cloudius commented:
> > I did not look at the patch series yet, but how is it different from getting all the mutations merged through a snapshot? Or does this series implements special handling for features into the snapshot transfer?
> 
> The list of features enabled at the time of the snapshot is provided in a separate field. The recipient checks whether it supports all of them before applying the mutations.

Yes. This is exactly what I was vehemently against of.
 
> 
> As far as I understand, the recipient shouldn't try to interpret mutations if it is not sure whether it will understand them. Sender doesn't have this problem and can compute the set of enabled features for the recipient.

So the features needs to be negotiated somewhere else. Before raft even enabled. Something that happens at random times and mostly out of our control is not a place for it.

> 
> > If yes, well here we see why it is a bad idea.
> 
> Could you elaborate? Is it because it complicates reasoning about whether it's OK to merge mutations? I don't think it is avoidable if we use group 0 to manage features as one command can affect how later commands are interpreted.
> 

Because the hack as we can see now propagates everywhere. And may position was that yes, group0 is not a pace to manage features. And I was told it will not be. It will be used as  a storage for them only, not for negotiation.


> Maybe we could have a different variant than `topology_change` just for features to make it more explicit, but it would definitely make the rest of the implementation more complicated. For example, when a node joins, it inserts basic information about itself to the system.topology table, including its own features and the join request - it should happen in a single command.

We should leave it to the gossiper or our new discovery protocol to negotiate features of a newly joining node.

kbr-scylla commented on test/topology/test_cluster_features.py:
```diff
@@ -0,0 +1,163 @@
+            assert TEST_FEATURE_NAME not in await get_enabled_features(cql, host)
+
+        await turn_on_test_feature_and_restart(manager, srv)
+        cql = await reconnect_driver(manager)
```
I don't understand why `reconnect_driver` is necessary; doesn't `wait_for_cql_and_get_hosts` do the job already?

piodul replied:
I might have misattributed some failure that I had when running the tests locally to scylladb/python-driver#170 and applied the `reconnect_driver` workaround. My installed version of the driver should not have this bug, and the tests don't seem to be flaky when I run them, so I'll drop the use of `reconnect_driver`.

kbr-scylla commented:
> So the features needs to be negotiated somewhere else. Before raft even enabled. Something that happens at random times and mostly out of our control is not a place for it.

What do you mean "random and mostly out of our control"? The check will happen at the initial read_barrier on bootstrap. If it fails we fail to join. If it succeeds we're good for the remaining time in our cluster. It's all good.

> We should leave it to the gossiper or our new discovery protocol to negotiate features of a newly joining node.

Using group 0 makes it easy to guarantee consistency. Other ad-hoc invented solutions won't make it simple. But why are we discussing this again? We discussed this on our daily calls and decided to go this way. You're the single person who doesn't like it.

gleb-cloudius commented:
> > So the features needs to be negotiated somewhere else. Before raft even enabled. Something that happens at random times and mostly out of our control is not a place for it.
> 
> What do you mean "random and mostly out of our control"? The check will happen at the initial read_barrier on bootstrap. If it fails we fail to join. If it succeeds we're good for the remaining time in our cluster. It's all good.

And hot is it apparent that raft read_barrier in fact used as a way to negotiate features. And it may happen before as well, because the leader pushes, so in fact you do not know at wat point after joining the group it happens.  I do not see how you can consider this as a good thing.

> 
> > We should leave it to the gossiper or our new discovery protocol to negotiate features of a newly joining node.
> 
> Using group 0 makes it easy to guarantee consistency. Other ad-hoc invented solutions won't make it simple. But why are we discussing this again? We discussed this on our daily calls and decided to go this way. You're the single person who doesn't like it.

There was a misunderstanding. Never ever I agreed on that and I do not remember we  decided to go this way despite my objection. The fiasco of this design leakage into the apply function just proves my point.

gleb-cloudius commented:
> > Because the hack as we can see now propagates everywhere.
> 
> What hack?
Negotiating features in snapshot transfer code.

> Propagates where?

Info apply function.

kbr-scylla commented:
> So the features needs to be negotiated somewhere else. Before raft even enabled. Something that happens at random times and mostly out of our control is not a place for it.

Please write down an alternative proposal.

We need to atomically do the following when joining the cluster:
- check if we support all enabled features
- if the check succeeds, add ourselves to the list that all other nodes will consider when calculating enabled features (so they don't enable a feature we don't support)

I don't think there's a simpler way than using `group0_guard` which allows us to do exactly that.
- take `guard`
- check if we support all features
- if the check succeeds, send command that adds ourselves and our supported features to cluster topology
- if the command succeeds the `state_id` check, we did this atomically and the cluster will not enable features without considering us

Why invent other ad-hoc solutions when we have nice existing framework to do these things?

kbr-scylla commented:
> Negotiating features in snapshot transfer code.

I see this more as a safety measure. The negotiation itself can happen explicitly on joining, as described above. But we need to handle the fact that once we join group 0 we start getting commands/snapshots from the leader, even before we do the explicit negotiation (which requires joining group 0). So protections like this should be in place.

gleb-cloudius commented:
> > So the features needs to be negotiated somewhere else. Before raft even enabled. Something that happens at random times and mostly out of our control is not a place for it.
> 
> Please write down an alternative proposal.

We were discussing the alternative proposals, but somehow abandoned the discussion and when for this (I was sure the discussion was postponed, not abandoned). One of the proposals was yours about letting the coordinator doing the check under the guard and adding the joining node if the check passes. What is wrong with that?

> 
> Why invent other ad-hoc solutions when we have nice existing framework to do these things?

This series implements ad-hoc solution. We need to find a proper one.

gleb-cloudius commented:
> > Negotiating features in snapshot transfer code.
> 
> I see this more as a safety measure. The negotiation itself can happen explicitly on joining, as described above.
Where?

> But we need to handle the fact that once we join group 0 we start getting commands/snapshots from the leader, even before we do the explicit negotiation (which requires joining group 0). So protections like this should be in place.

We need to join only when we know we support all needed features. After that what protection do you need exactly? You support all the features and you need to be able to handle them no matter if they enabled or not.

kbr-scylla commented:
> One of the proposals was yours about letting the coordinator doing the check under the guard and adding the joining node if the check passes. What is wrong with that?

That would be exactly like this PR, except the check+add would happen on coordinator instead of the joining node. From correctness POV there's no difference, only adds more boilerplate: you will need a separate RPC to join and so on.

We would still be updating supported features stored in group 0 when restarting. We would still be enabling features by calculating supported feature intersection in `group0_state_machine::apply()`. We would still do the sanity check that we don't try to enable some illegal feature, only those we support.

kbr-scylla commented:
> Why not calculate on the coordinator and let it enable features?

That's a complication. It would require additional storage and separate command to enable a feature, apart from the existing storage and command to support a feature by a node (so that nodes learn that the coordinator decided to enable a feature). The calculation of enabled features from stored supported features is deterministic, so there's no reason to add additional state. Generally if we have states `X, Y, Z` and `Z = f(X, Y)` where `f` is a function, `Z` is redundant and we can remove it unless it's an important oprimization (e.g. `f` is hard to calculate - not the case here).

kbr-scylla commented:
> You do not need new commands, just one more field "enabled_futures" (don't we have one already today). When the field changes features are enabled. Calculating it on a coordinator is a simplification. Only coordinator knows that new node is not joining in.

The field is redundant. If every member of the topology supports a feature, it is safe for a node to consider it enabled right away. There is no value in having a separate step that the coordinator has to perform. It doesn't matter if a node is joining or not. If the joining node supports the feature and all other nodes support the feature then we can consider it enabled. If the joining node doesn't support the feature then we don't enable it. It's the same as for all other nodes, no need to treat the joining node differently. Just ensure that every member of topology fills the `supported_features` field correctly which we need to do in any solution. Treating joining nodes differently is a complication. Having a separate redundant step is a complication.

xemul commented on gms/feature_service.cc:
```diff
@@ -249,10 +249,8 @@ future<> feature_service::enable_features_on_startup(db::system_keyspace& sys_ks
         // enabled in the code by default, so just skip it.
     }
 
-    co_await container().invoke_on_all([&features_to_enable] (auto& srv) -> future<> {
```
I don't get the patch title. It says "make it enable features on all shards", but current code _already_ _does_ _so_, doesn't it?

piodul replied:
The comment was about the "feature_service::enable" function. It didn't enable features on all shards _by itself_, and it had to be called via `invoke_on_all` to do that.

I did it mainly for consistency with the second, "peristent" overload of that function that I introduced in later commits.

xemul commented on gms/feature_service.hh:
```diff
@@ -64,6 +64,7 @@ public:
     explicit feature_service(feature_config cfg);
     ~feature_service() = default;
     future<> stop();
+    future<> enable(std::set<std::string_view> list, db::system_keyspace& sys_ks);
```
I don't see the need to tune feature_service API to allow for persistent and non-persistent enabling. Current code allows you both already. If you need to enable features without persisting, use feature_service::enable(), if you need to persist enabling -- get the persistent enabler and use _it_.

piodul replied:
@xemul I needed to expose the possibility of enabling features persistently from outside (from raft group 0). I don't see how to cleanly expose the persistent enabler, could you tell what you had in mind?

The persistent enabler seems to be designed to be a listener, perhaps I could add a way to register callbacks that listen to changes to the group 0 state. I'm not sure if I like that idea, I think features should be enabled explicitly there.

Besides, I think that the current `feature_service::enable` API is a bit unsafe. Enabling features without persisting them first is not safe in general, and the fact that this method is called `feature_service::enable` and is public makes it too tempting to use it. I think the API should be more explicit about what it does, which is what I tried to do with this cleanup.

(Another besides: I'll move the changes to the feature service to a separate PR, but let's first clear the issues here first)

piodul replied:
@xemul ping

xemul replied:
Sorry

> @xemul I needed to expose the possibility of enabling features persistently from outside (from raft group 0). I don't see how to cleanly expose the persistent enabler, could you tell what you had in mind?

In that case you need to patch persistent enabler, not feature service. The latter is nowadays a "transport" that spreads features enabling around subscribers

> The persistent enabler seems to be designed to be a listener, perhaps I could add a way to register callbacks that listen to changes to the group 0 state. I'm not sure if I like that idea, I think features should be enabled explicitly there.

Not that it's _designed_ to be a listener. There was just one source of features-to-be-enabled and it was gossiper. Now it's raft, so feel free to patch this enabler respectively

> Besides, I think that the current feature_service::enable API is a bit unsafe. Enabling features without persisting them first is not safe in general, and the fact that this method is called feature_service::enable and is public makes it too tempting to use it. I think the API should be more explicit about what it does, which is what I tried to do with this cleanup.

It's not supposed to be safe, because, as I mentioned above, feature_service is just a way to tell the rest of the system about enabling features. It's been such in the old days, then it was considered that not persisting features was unsafe (a2f5ad432f657253a3a9475be152ee52c05ce30f) and it was done by patching feature-service introducing nasty dependency loops via ugly `qctx` thing. I propose not to walk this route for the 2nd time

> (Another besides: I'll move the changes to the feature service to a separate PR, but let's first clear the issues here first)

:+1:

xemul commented on gms/feature_service.cc:
```diff
@@ -266,11 +266,18 @@ future<> feature_service::enable(std::set<std::string_view> features, db::system
+    bool any_new_features_to_enable = false;
     for (feature& f : registered_features() | boost::adaptors::map_values) {
         if (!f && features.contains(f.name())) {
+            any_new_features_to_enable |= !feats_set.contains(f.name());
```
The std::set::emplace() returns boolean whether the insertion took place, you can rely on it to update the `any_new_features_to_enable`

gleb-cloudius commented:
> > You do not need new commands, just one more field "enabled_futures" (don't we have one already today). When the field changes features are enabled. Calculating it on a coordinator is a simplification. Only coordinator knows that new node is not joining in.
> 
> The field is redundant. If every member of the topology supports a feature, it is safe for a node to consider it enabled right away.

Not unless a coordinator already decided to join a node with some, not yet enabled futures, missing. Actually doing it through the coordinator will allow us to decide when we want to enable the future after upgrade.

> There is no value in having a separate step that the coordinator has to perform. It doesn't matter if a node is joining or not. If the joining node supports the feature and all other nodes support the feature then we can consider it enabled. If the joining node doesn't support the feature then we don't enable it.  It's the same as for all other nodes, no need to treat the joining node differently. Just ensure that every member of topology fills the `supported_features` field correctly which we need to do in any solution. Treating joining nodes differently is a complication. Having a separate redundant step is a complication.

I do not understand what is your argument. Can you delegate it to nodes without adding hacks into apply/snapshot transfer path. Fine, do it. If you cannot then I show you the way how you can and then you do not have an argument.

gleb-cloudius commented:
> Anyway if your main objection is needing the safety checks inside command/snapshot application, then indeed adding the node by the coordinator would be a solution (we could remove the safety checks and trust the coordinator). No need for separate `enabled_futures` field for that.

If this is the case then perfect! We already trust the coordinator.

xemul commented on gms/feature_service.hh:
```diff
@@ -58,6 +60,9 @@ class feature_service final : public peering_sharded_service<feature_service> {
 
     feature_config _config;
+
+    shared_ptr<persistent_feature_enabler> _enabler;
```
Please, no. The persistent_feature_enabler already references feature_service, this back-link is asking for troubles. But this comment is connected to my other one regarding the controversial feature_service API change.

kbr-scylla commented:
> Can you delegate it to nodes without adding hacks into apply/snapshot transfer path.

BTW. you still need "hacks".

Suppose that we do exactly as you say: coordinator joins the node, and we even add separate `enabled_features` field, and make only the coordinator update this field.

You **still** need to make sure that you don't merge the command which enables this feature with a later command that depends on it. Because any replica when it applies the second command must have applied the earlier command first.

The exact same thing happens in snapshot transfer. You **still** need to  locally enable all the coordinator-enabled features before you apply the rest of the snapshot.

gleb-cloudius commented:
> You **still** need to make sure that you don't merge the command which enables this feature with a later command that depends on it. Because any replica when it applies the second command must have applied the earlier command first.

We discussed that on the doc. A node should be able to process a command that depends on a feature it supports even if the node itself did not enable it yet. Otherwise today feature enablement would not have worked as well since not all nodes enabled the future at the same time.
    
IOW a node start to accept new features the moment it advertises it as "supported". It starts produce output as if new feature is enabled when it sees that the entire cluster supports it.

kbr-scylla commented:
> We discussed that on the doc. A node should be able to process a command that depends on a feature it supports even if the node itself did not enable it yet. Otherwise today feature enablement would not have worked as well since not all nodes enabled the future at the same time.

Hmmm....
now I wonder if it actually works today.

Because we do have the concept of enabling a feature locally. We even persist the fact that a feature is enabled locally on each node. And I guess that this **is** used for something. That is, if a feature is even supported but not enabled, we make different decisions, no?

For example, wasn't there something with truncating mutations? That if we consider some feature disabled, we process mutations that depend somehow on this feature differently.

But then that would mean the entire thing is incorrect, as you say - because a node starts producing output based on a feature as soon as it observes that everybody supports it. Then another node, which didn't yet observe this, obtains this output and makes a wrong decision.

kbr-scylla commented:
@tgrabiec could you comment on this? https://github.com/scylladb/scylladb/pull/13497#issuecomment-1563972667

gleb-cloudius commented:
> > We discussed that on the doc. A node should be able to process a command that depends on a feature it supports even if the node itself did not enable it yet. Otherwise today feature enablement would not have worked as well since not all nodes enabled the future at the same time.
> 
> Hmmm.... now I wonder if it actually works today.

Why not?

> 
> Because we do have the concept of enabling a feature locally. We even persist the fact that a feature is enabled locally on each node. And I guess that this **is** used for something. I.e. if a feature is even supported but not enabled, we make different decisions, no?

We started to persist them pretty recently (https://github.com/scylladb/scylladb/commit/0df14a48cf1dec07df5c7530f6cc84cbb8a49600) . The patch set claimed to be part of the solution for https://github.com/scylladb/scylladb/issues/4458, but I do not think other part (which one) was ever implemented since the bug is still open. And frankly the solution for the issue is to use raft for topology change all along. 

I do see at least one good reason to persist enabled features if we read it early enough: we need to know latest sstable format when replaying commitlog.

> 
> For example, wasn't there something with truncating mutations? That if we consider some feature disabled, we process mutations that depend somehow on this feature differently.

Don't remember, but this would be a very bad designed feature.

piodul commented:
> That would be exactly like this PR, except the check+add would happen on coordinator instead of the joining node. From correctness POV there's no difference, only adds more boilerplate: you will need a separate RPC to join and so on.

Yes... We will have to introduce an RPC that inserts a row to the system.topology table on behalf of the joining node, probably in some new "pre_join" state, so that the node's features are accounted for. We will have to do something similar on downgrade because it won't be safe to start group 0 server until we have a guarantee that the node supports all enabled features.

I started exploring this approach previously but didn't like the complexity incurred by the additional states, RPCs, etc. I can revisit it as there are objections to the current one.

Though, I do not fully understand the arguments behind the objections. Yes, feature check via group 0 barrier is not explicit, but on the other hand doing everything inside group 0 looks cleaner to me because we don't have to invent additional states or ad-hoc RPCs. By doing the checks on apply/snapshot transfer we just guarantee that a node won't apply something that it doesn't understand. Do a read barrier, and if it succeeds try to insert the features - as simple as that. If it fails then we mustn't continue booting anyway. If we succeed, then we are good because our features are accounted for in group 0 state - at least until we leave the cluster.

I also did not understand the comment about this "hack propagating everywhere". I understand that the feature check there is very non-standard use, but I don't see how it breaks things from the architecture perspective (I suppose this was the meaning) or affects the rest of the group 0.

kbr-scylla commented:
> Though, I do not fully understand the arguments behind the objections. Yes, feature check via group 0 barrier is not explicit, but on the other hand doing everything inside group 0 looks cleaner to me because we don't have to invent additional states or ad-hoc RPCs.

I agree.

gleb-cloudius commented:
> 
> Though, I do not fully understand the arguments behind the objections. Yes, feature check via group 0 barrier is not explicit, but on the other hand doing everything inside group 0 looks cleaner to me because we don't have to invent additional states or ad-hoc RPCs. By doing the checks on apply/snapshot transfer we just guarantee that a node won't apply something that it doesn't understand. Do a read barrier, and if it succeeds try to insert the features - as simple as that. If it fails then we mustn't continue booting anyway. If we succeed, then we are good because our features are accounted for in group 0 state - at least until we leave the cluster.

Sorry, but I do not see how "lets join the cluster without feature check and see if it works (and if it does not - who cares we should not have joined anyway)" approach can be seriously considered.

> 
> I also did not understand the comment about this "hack propagating everywhere". 

It has implications where it should not have one. Now command merging suddenly needs to take it into account. And I am sure in the future it will stand in our way in ways we cannot imagine now.

kbr-scylla commented:
> lets join the cluster without feature check

That's not what the PR does. The feature check is done inside `apply` and `transfer_snapshot`.

gleb-cloudius commented:
> > Though, I do not fully understand the arguments behind the objections. Yes, feature check via group 0 barrier is not explicit, but on the other hand doing everything inside group 0 looks cleaner to me because we don't have to invent additional states or ad-hoc RPCs.
> 
> I agree.

The is a difference between "doing everything inside group 0" which we obviously will do even if the check will be done on a topology coordinator and hiding it in a group0 state machine plumbing in a way that the error is reported by spilling the shit because of a clog (well, we shouldn't have used the un-compatible toilet in the first place). Sorry for the graphic analogy.

gleb-cloudius commented:
> > lets join the cluster without feature check
> 
> That's not what the PR does. The feature check is done inside `apply` and `transfer_snapshot`.

`apply` and `transfer_snapshot` are called only after the node joins group0, so how is it not what PR does? I am confused.

kbr-scylla commented:
> apply and transfer_snapshot are called only after the node joins group0, so how is it not what PR does? I am confused.

You said "join the cluster", not "join group 0".

gleb-cloudius commented:
> > apply and transfer_snapshot are called only after the node joins group0, so how is it not what PR does? I am confused.
> 
> You said "join the cluster", not "join group 0".

I mean raft cluster here, sorry for the confusion.

I cannot wrap my head about the proposed way to check features by  joining a raft cluster and see if it is usable. I just can't.

gleb-cloudius commented:
> The feature check is done inside `apply` and `transfer_snapshot`.

Also why do it on each `apply` and `transfer_snapshot` if it is needed only once?

kbr-scylla commented:
> I cannot wrap my head around the proposed way to check features by joining a raft cluster and see if it is usable. I just can't.

It's similar to using gossiper to check features. We start gossiping with other nodes to obtain the features list, then we check features and leave if needed.

Same with group 0, we join as a non-voter and passive participant - read-only, we don't send any commands on our own. Then we check features and leave if needed.

> Also why do it on each apply and transfer_snapshot if it is needed only once?

It actually isn't done only once, we may have to do it at least on every restart when we could downgrade and attempt unsupporting an alraedy-enabled features. How do we solve this problem within your proposed framework?

gleb-cloudius commented:
> > I cannot wrap my head around the proposed way to check features by joining a raft cluster and see if it is usable. I just can't.
> 
> It's similar to using gossiper to check features. We start gossiping with other nodes to obtain the features list, then we check features and leave if needed.

No, it is not similar. Shadow round was invented for exactly such cases where we can get the info about a cluster without actually joining. With group0 we can do even better. Just ask any node that already in the cluster to do the check and join you.

> 
> Same with group 0, we join as a non-voter and passive participant - read-only, we don't send any commands on our own. Then we check features and leave if needed.
> 

You check it inside the raft plumbing while the raft cluster you joined is actually not fully functional since a leader tries to push to a node that constantly rejects messages.
 
> > Also why do it on each apply and transfer_snapshot if it is needed only once?
> 
> It actually isn't done only once, we may have to do it at least on every restart 

This is still far from doing it on each message.

> when we could downgrade and attempt unsupporting an alraedy-enabled features. How do we solve this problem within your proposed framework?

Check against enabled features on restart?

kbr-scylla commented:
> Check against enabled features on restart?

But we are already a member of group 0 - which, if I understand your arguments, is a problem (hence you want to check features before joining).

So we'd have to check it before group 0 is enabled on restart. Is there a way to do this consistently @piodul? For example, will comparing to our pre-restart group 0 state (which doesn't have to be the newest - we do this before read barrier) be good enough?

piodul commented:
> So we'd have to check it before group 0 is enabled on restart. Is there a way to do this consistently @piodul? For example, will comparing to our pre-restart group 0 state (which doesn't have to be the newest - we do this before read barrier) be good enough?

It's not good enough. The problem is that, while a node is down, some other node might be upgraded and some feature becomes enabled. The node that downgrades would have to ask some other, existing node in the cluster to do the feature check and update, like when joining.

> If we will delegate feature enablement to the coordinator the coordinator may do a global barrier to make sure all nodes knows each other feature set before enabling a feature. In this case on reboot the node needs to check that if it lacks previously present feature the feature is also not yet available on some of the other nodes (and hence cannot be enabled yet).

That sounds like an alternative, but introduces a constraint: now features cannot be enabled if some node is down. The current design (or the one that uses RPC on join+downgrade) only requires quorum.

gleb-cloudius commented:
> > If we will delegate feature enablement to the coordinator the coordinator may do a global barrier to make sure all nodes knows each other feature set before enabling a feature. In this case on reboot the node needs to check that if it lacks previously present feature the feature is also not yet available on some of the other nodes (and hence cannot be enabled yet).
> 
> That sounds like an alternative, but introduces a constraint: now features cannot be enabled if some node is down. The current design (or the one that uses RPC on join+downgrade) only requires quorum.

I do not think it is unreasonable constraint. Feature enablement  usually happens  during cluster upgrade that already requires all nodes to be alive (you cannot upgrade a dead node) and if one or more nodes are dying during rolling upgrade you probably want to downgrade anyway, so not enabling a feature is a feature (no pan intended) not a bug in this case.

kostja commented on db/system_keyspace.cc:
```diff
@@ -256,6 +256,7 @@ schema_ptr system_keyspace::topology() {
             .with_column("num_tokens", int32_type)
             .with_column("shard_count", int32_type)
             .with_column("ignore_msb", int32_type)
+            .with_column("supported_features", utf8_type)
```
@kbr-scylla I think it goes hand in hand how CQL statements are built to enable/disable features. If we use set<text> and use + operator to add features to the set, we can make the end result eventually consistent, so there is an advantage. There is also a guarantee that the feature set only contains unique features at all steps. Looks like a separate follow up work, since for the transitioning commit as it is now, it is easier to *move* features to raft as is, and then  change the way htey are represented in raft. I am concerned with the scope of this series.

avikivity replied:
Eventually consistent? Isn't it managed by Raft?

I don't think there's a difference in terms of correctness (since we know how to implement sets in C++), but it's much nicer to have a set. It's also better to start with a set, incremental development and schema changes don't go together well.

kostja commented on service/storage_service.cc:
```diff
@@ -1520,7 +1524,8 @@ future<> storage_service::update_topology_with_local_metadata(raft::server& raft
                .set("ignore_msb", local_ignore_msb)
-               .set("release_version", local_release_version);
+               .set("release_version", local_release_version)
+               .set("supported_features", gms::feature_service::from_feature_set(local_supported_features));
```
The update happens under group0 client already., which ensures the read barrier. All we need to do is that supported_feature_set() from feature_service is a superset of the persisted state here.

kostja commented:
The joining of Raft cluster doesn't happen implicitly, someone has to submit set_configuration() command to introduce the new node. Historically, before we introduced the centralized topology coordinator it was any node, which acted on behalf of the joining node. We even implemented command forwarding to do this transparently from any existing cluster node. That node handles a special RPC we introduced - GROUP0_MODIFY_CONFIG, which is Scylla-level RPC command, which then can execute Raft-level command to modify config. 

Let me re-iterate, that this architecture with transparent forwarding was designed in pre-coordinator era, when every cluster node acted, essentially, on behalf of the coordinator. With features in place, we have the following options options:
- ensure that the centralized coordinator is always the target for GROUP0_MODIFY_CONFIG RPC. Then it can take a read barrier and validate that the joining node supports the right feature set. To do that, we will need to change discovery to always return the coordinator ID as raft leader id (which it already does, but we don't bounce the request on a non-leader).
- perform the check on the interim node, acting on behalf of the joining node. The interim node will need to protect the change with state id.
- pefrom the check on the joining node itself, by having discovery return the relevant feature set *and* state id.

There is also an option to do the check during the gossip shadow round, but as has been discussed at length it is race-prone.
Whatever option we choose it can be done in a separate follow up patch.

kostja commented:
@gleb-cloudius unless we modify GROUP0_MODIFY_CONFIG to do a read barrier and perform a feature check, we have a chicken-and-egg problem. 
- By submitting MODIFY_CONFIG command, the node *promises* that it can take a snapshot
- The promise to take a snapshot is a promise that it supports the feature set used to create it
- This feature set is in the snapshot.

We could peek into the snapshot (which you don't like) and allow the node to violate its promise (which you don't like). Then the centralized coordinator will have to clean up after such faulty node (which you also don't like). So to avoid things you don't like (indeed because they are hard to reason about and hard to get right), the suggestion is to do the check in GROUP0_MODIFY_CONFIG and make it validate the feature set on behalf of the joining node. Then the joining node will not be making promises it cannot keep.

gleb-cloudius commented:
> @gleb-cloudius unless we modify GROUP0_MODIFY_CONFIG to do a read barrier and perform a feature check, we have a chicken-and-egg problem.

We replace the rpc with the different one that will do some checks but some will be delegated to the leader. See

https://docs.google.com/document/d/1iQH7RzYDaNHrAAoi0JmWdJGe9OYtZKx_oYHJerPN-FM/edit?usp=sharing

avikivity commented:
Pre-review review:
 - schema changes must be documented
 - cover letter should explain upgrade/downgrade strategy
     -  old version -> new version
     -  new version (with raft group0 disabled) -> new version (with raft group0 enabled)
     - combined?

avikivity commented on gms/feature_service.cc:
```diff
@@ -257,19 +257,24 @@ future<> persistent_feature_enabler::enable_features() {
 
     std::set<std::string_view> features_v = boost::copy_range<std::set<std::string_view>>(features);
-    co_await _feat.enable(std::move(features_v), feature_service::without_persisting{});
+    co_await enable(std::move(features_v), feature_service::without_persisting{});
```
@xemul, thinking with modularity hat on:

feature_service should expose a feature_service_persistence_interface
main should implement feature_service_persistence_interface with system_keyspace
cql_test_env can implement feature_service_persistence_interface with std::unordered_set

xemul replied:
@avikivity ,  `gms::persistent_feature_enabler` can already implement feature_service_persistence_interface with system_keyspace , but it also needs gossiper. The question is where features-to-enable come from

avikivity replied:
We can go wild and have every control-plane module declare its interfaces rather than talk to other control-plane modules directly.

However, without the build system enforcing isolation, we'll keep slipping and re-adding dependencies.

I'll send out something shortly.

Also cc @tchaikov who is relevant for modules, cmake, and modularity.

kostja commented on service/topology_state_machine.hh:
```diff
@@ -111,6 +111,11 @@ struct topology {
+    // Checks whether the node described by `local_features` supports all
+    // enabled features described by `enabled_features`. Throws an exception
+    // if it doesn't.
+    static void check_knows_enabled_features(const std::set<std::string_view>& local_features, const std::set<std::string_view>& enabled_features);
```
One reason why enabled features should be persistent is that check_knows_enabled_features() here behavior is predictable. Imagine there is a node which joined and then left the cluster. Then check_knows() will be returning a different value for the same feature set. Of course the feature set may change over time, and then check_knows_enabled_features() also starts returning a new value. But the set of features should only grow, never shrink. In other words, to make it easy to reason about the behaviour of check_knows_enabled_features() between restarts and with interim nodes joining and leaving the clusters, the set of enabled features should be persisted.

kostja commented on idl/storage_service.idl.hh:
```diff
@@ -27,6 +27,7 @@ namespace service {
   struct raft_topology_snapshot {
       std::vector<canonical_mutation> topology_mutations;
       std::optional<canonical_mutation> cdc_generation_mutation;
+      std::vector<sstring> enabled_features [[version 5.4]];
```
If the set of features is persisted in the topology table, it is automatically part of the snapshot, so this patch is unnecessary.

kostja commented on service/storage_service.cc:
```diff
@@ -317,6 +317,20 @@ future<> storage_service::topology_state_load(cdc::generation_service& cdc_gen_s
+    //
+    // An interesting and useful side effect of those checks is that
+    // performing a successful read barrier makes sure that the local node
+    // supports all cluster's features.
```
I would appreciate refactoring of the topology_state_load() and splitting it to load() and reload() where reload() automatically detects the delta in the topology change and only invokes state change triggers if the delta is present. It will make much easier to reason that various state change triggers have reasonable complexity, e.g. in this specific case it would not matter whether or not enable() is a no-op when features don't change, because it won't be invoked if features don't change. Not part of this patch scope.

kostja commented on gms/feature_service.cc:
```diff
@@ -305,4 +308,12 @@ future<> feature_service::enable(std::set<std::string_view> list, feature_servic
 }
 
+future<> feature_service::disconnect_from_gossip(gossiper& g) {
+    _connected_to_gossip = false;
```
We should look at upgrade state here indeed. Later on we can just remove it.

piodul commented:
The tests part have been improved upon and re-submitted separately here: https://github.com/scylladb/scylladb/pull/14211

piodul commented:
More has been submitted here: https://github.com/scylladb/scylladb/pull/14722

piodul commented:
Closing - the new implementation is going to be different, and some parts of it were already mentioned here and merged. No need to keep this open.
